modules/siddhi-launcher/src/main/java/org/wso2/siddhi/sdk/launcher/debug/internal/DebugRuntime.java

```diff
@@ -0,0 +1,158 @@
+import org.wso2.siddhi.sdk.launcher.debug.BreakPointInfo;
+import org.wso2.siddhi.sdk.launcher.debug.VMDebugManager;
+import org.wso2.siddhi.sdk.launcher.exception.InvalidExecutionStateException;
+import org.wso2.siddhi.sdk.launcher.exception.NoSuchStreamException;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The DebugRuntime which handles the siddhi debugger for each siddhi app.
+ */
+public class DebugRuntime {
+
+    private static final Logger log = Logger.getLogger(DebugRuntime.class);
+
+    private Mode mode = Mode.STOP;
+    private transient String siddhiApp;
+    private transient String siddhiAppFileName;
+    private transient SiddhiAppRuntime siddhiAppRuntime;
+    private transient SiddhiDebugger debugger;
+
+    public DebugRuntime(String siddhiAppFileName, String siddhiApp) {
+        this.siddhiApp = siddhiApp;
+        this.siddhiAppFileName = siddhiAppFileName;
+        createRuntime();
+    }
+
+    public SiddhiAppRuntime getSiddhiAppRuntime() {
+        return siddhiAppRuntime;
+    }
+
+    private void setSiddhiAppRuntime(SiddhiAppRuntime siddhiAppRuntime) {
+        this.siddhiAppRuntime = siddhiAppRuntime;
+    }
+
+    public String getSiddhiAppFileName() {
+        return siddhiAppFileName;
+    }
+
+    public SiddhiDebugger getDebugger() {
+        return debugger;
+    }
+
+    public void debug() {
+        if (Mode.STOP.equals(mode)) {
+            debugger = siddhiAppRuntime.debug();
+            debugger.setDebuggerCallback((event, queryName, queryTerminal, debugger) -> {
+                String[] queries = getQueries().toArray(new String[getQueries().size()]);
+                int queryIndex = Arrays.asList(queries).indexOf(queryName);
+                //Sending message to client on debug hit
+                log.info("@Debug: Query: " + queryName + ", Terminal: " + queryTerminal + ", Event: " +
+                        event);
+                Map<String, Object> queryState = this.debugger.getQueryState(queryName);
+                BreakPointInfo breakPointInfo = new BreakPointInfo(siddhiAppFileName, queryIndex, queryTerminal
+                        .toString());
+                breakPointInfo.setQueryState(queryState);
+                breakPointInfo.setQueryName(queryName);
+                breakPointInfo.setEventInfo(event);
+                VMDebugManager.getInstance().getDebugSession().notifyHalt(breakPointInfo);
+            });
+            mode = Mode.DEBUG;
+        } else if (Mode.FAULTY.equals(mode)) {
+            throw new InvalidExecutionStateException("Siddhi App is in faulty state.");
+        } else {
+            throw new InvalidExecutionStateException("Siddhi App is already running.");
+        }
+    }
+
+    public void stop() {
+        if (debugger != null) {
+            debugger.releaseAllBreakPoints();
+            debugger.play();
+            debugger = null;
```

Comment:
I do not see the need to assign null to this or SiddhiAppRuntime

Reply:
If the stop() called twice then there can be an error occurred. To avoid that once the stop() is called I assign null to it.